Initial commit for EOA improvement RIP-7560 transaction subtypes (#1)
eth-infinitism/RIPs: RIPS/rip-9999.md

to existing EOA functionality.
These "EOA extensions" can either be seen as a step towards "Full Native Account
Abstraction", or be evaluated on their own merit regardless of the eventual Account Abstraction design.

This proposal is intended to prevent the fragmentation and inconsistencies that may arise as a result of implementing
multiple independent standalone proposals by combining all of them into a unified design.

Here is the list of EIPs that propose adding features to EOA accounts that are addressed by this document:

* [EIP-2711: Sponsored, expiring and batch transactions](https://eips.ethereum.org/EIPS/eip-2711)
* [EIP-3074: AUTH and AUTHCALL opcodes](https://eips.ethereum.org/EIPS/eip-3074)
* [EIP-5003: Insert Code into EOAs with AUTHUSURP](https://eips.ethereum.org/EIPS/eip-5003)
* [EIP-5081: Expirable Transaction](https://eips.ethereum.org/EIPS/eip-5081)
* [EIP-5806: Delegate transaction](https://eips.ethereum.org/EIPS/eip-5806)
* [EIP-7377: Migration Transaction](https://eips.ethereum.org/EIPS/eip-7377)

Each one of these proposals introduces features that can benefit both existing users and potential future users of
the Ethereum blockchain in multiple ways.

> However, together these proposals consist 3 different transaction types and 3 opcodes, which is probably unsustainable.

Reviewer: 3 different transaction types, which means they are incompatible with each other...

Reviewer 2: [](url)

[ERC-1613: Gas Stations Network](https://eips.ethereum.org/EIPS/eip-1613)
before that.

None of these proposals fully support validity time ranges, which enable both scheduled and expiring transactions,
adding a noticeable improvement to end-user experience with a relatively small amount of effort.

And most importantly, none of these proposals form a part of a bigger vision.
Instead, these proposals are isolated patches that can be applied in a single design iteration.
These proposals cannibalize each other's use-cases, but do not fully resolve the pain points Ethereum users are
currently suffering from, and they do not fully align with any kind of vision for a Full Native Account Abstraction.

## Specification

### Shared RIP-9999 transaction payload

The required synergy of the proposed transaction subtypes is achieved by defining a shared base structure of
a transaction.
This will ensure the only differences between these subtypes are the ones strictly required
by their semantics instead of being caused by different tastes or variations in implementation details.

These are the shared parameters of any RIP-9999 transaction subtype payload:

```
  chainId,
  nonce,
  callData,

>   paymasterData,

Reviewer: missing `paymaster`

Reviewer 3: probably because RIP-7560 also doesn't separate `paymaster` from `paymasterData`, for ERC-4337 compatibility.

Reviewer: AFAIK, this RIP defines the transaction, which is the API type. In the ABI-packed version we pack them together.
in a transaction type, we can also have optional items.
also, the paymasterGasLimit (and the missing postOpGasLimit) are also part of the paymasterAndData

  paymasterGasLimit,
  maxPriorityFeePerGas,
  maxFeePerGas,
  validAfter,
  validUntil,
  accessList,

>   yParity, r, s

Reviewer: for future-compatibility, better add `signature` field, which is currently parsed as yParity,r,s

Author: Moved the `yParity, r, s` to specific subtypes. It seems like all current EIPs use "yParity, r, s" for transaction type signature, and as this document targets EOAs explicitly this is acceptable.

[ERC-4337: Account Abstraction Using Alt Mempool](https://eips.ethereum.org/EIPS/eip-4337)
and even
[ERC-1613: Gas Stations Network](https://eips.ethereum.org/EIPS/eip-1613)
before that.

None of these proposals fully support validity time ranges, which enable both scheduled and expiring transactions,
adding a noticeable improvement to end-user experience with a relatively small amount of effort.

And most importantly, none of these proposals form a part of a bigger vision.
Instead, these proposals are isolated patches that can be applied in a single design iteration.
These proposals cannibalize each other's use-cases, but do not fully resolve the pain points Ethereum users are
currently suffering from, and they do not fully align with any kind of vision for a Full Native Account Abstraction.

## Specification
The aforementioned EIPs can be implemented by the following transaction subtypes:

### Shared RIP-9999 transaction payload

The required synergy of the proposed transaction subtypes is achieved by defining a shared base structure of
a transaction.
This will ensure the only differences between these subtypes are the ones strictly required

> by their semantics instead of being caused by different tastes or variations in implementation details.

Reviewer 4: Why create a new tx type within a type? Clients will need to now parse out the 2718 tx type and then do a match on the number of RLP elements in the list to figure out what subtype it is. This implicitly means that subtypes cannot have the same length of parameters, otherwise it will not be possible to determine which tx type to decode to.

Author: This RIP does propose a set of potential transaction types but we wanted to avoid hogging the entire types-space with our proposals, and these types share a lot in common.
So the idea behind defining a "subtype" if basically to extend the EIP-2718 type field to be 16 bit. This allows us to create 256 "subtypes" within the "Type 4" transaction without a risk interfering with any other EIP or RIP. Clients will need to parse out the EIP-2718 transaction type and, if it equals 4, also parse out the AA transaction subtype.

Something similar will probably need to happen in Solidity, where specific structs will need to be strictly extending the base transaction struct. This part is still missing from the document.

Does this approach make sense to you? Do you have a suggestion on how we can address this issue better?
Thanks!

Reviewer 4: Since there are still ~120 available EIP-2718 tx types I think you should use them directly. The subtype adds a lot of complexity to clients and tooling. If we ever get into a situation where we're nearing the end of the available range, we can institute subtype format.

  nonce,
  callData,
  paymasterData,
  paymasterGasLimit,
  maxPriorityFeePerGas,
  maxFeePerGas,
  validAfter,
  validUntil,
  accessList
```

### Gas-sponsored time-limited EOA transaction subtype

```
SPONSORED_EOA_SUBTYPE = x
```

This RIP-7560 transaction subtype allows ECDSA signature and nonce validation, same as the one used in all
of existing transaction types, to be applied to a transaction that can also provide `paymasterData`, `validAfter`
and `validUntil` parameters.

The transaction payload for this subtype is extended with the following fields:
```
 to
 value
 callGasLimit,
 yParity, r, s
```

This subtype extends the ability of an RIP-7560 transaction to specify a Paymaster contract, 
which accepts the gas charges, to all EOA accounts.
It also provided an EOA with an ability to explicitly specify a time range where this
transaction is valid, creating both "expiring" and "delayed" transactions.

In case the `to` field is `nil`, `callData` is treated as `initCode` for the purpose of contract creation,
enabling Gas Abstracted new smart contract deployment.

### Gas-sponsored time-limited batched execution EOA transaction subtype

```
BATCHED_EOA_SUBTYPE = x
```

> This transaction subtype extends the `SPONSORED_EOA_SUBTYPE` transactions with an ability to initialize multiple

Reviewer 5: If it extends the previous subtype, it's implied that it also has the same fields:
 to
 value
 callGasLimit,
 yParity, r, s
 
 But it doesn't seem the case. They seem contradictory.

Author: Removed the "chaining", is this better?

https://github.com/eth-infinitism/RIPs/pull/1/commits/cb023019f2a1d75a46e4d97128af276c1e80b57c

Abstraction", or be evaluated on their own merit regardless of the eventual Account Abstraction design.

This proposal is intended to prevent the fragmentation and inconsistencies that may arise as a result of implementing
multiple independent standalone proposals by combining all of them into a unified design.

Here is the list of EIPs that propose adding features to EOA accounts that are addressed by this document:

* [EIP-2711: Sponsored, expiring and batch transactions](https://eips.ethereum.org/EIPS/eip-2711)
* [EIP-3074: AUTH and AUTHCALL opcodes](https://eips.ethereum.org/EIPS/eip-3074)
* [EIP-5003: Insert Code into EOAs with AUTHUSURP](https://eips.ethereum.org/EIPS/eip-5003)
* [EIP-5081: Expirable Transaction](https://eips.ethereum.org/EIPS/eip-5081)
* [EIP-5806: Delegate transaction](https://eips.ethereum.org/EIPS/eip-5806)
* [EIP-7377: Migration Transaction](https://eips.ethereum.org/EIPS/eip-7377)
* [EIP-7553: Separated Payer Transaction](https://eips.ethereum.org/EIPS/eip-7553)

Each one of these proposals introduces features that can benefit both existing users and potential future users of
the Ethereum blockchain in multiple ways.

> However, together these proposals consist 3 different transaction types and 3 opcodes, which is probably unsustainable.

Reviewer 5: ```suggestion
However, together these proposals consist of 3 different transaction types and 3 opcodes, which is probably unsustainable.
```

None of these proposals supports Gas Abstraction mechanism that would be conceptually similar to "Paymasters" - smart
contracts capable of paying transaction gas, based on some on-chain validity function, but do not gain any
privileged access to sender's funds.
This is very disappointing given that the concept of "Paymasters" has being widely used in production by
[ERC-4337: Account Abstraction Using Alt Mempool](https://eips.ethereum.org/EIPS/eip-4337)
and even
[ERC-1613: Gas Stations Network](https://eips.ethereum.org/EIPS/eip-1613)
before that.

None of these proposals fully support validity time ranges, which enable both scheduled and expiring transactions,
adding a noticeable improvement to end-user experience with a relatively small amount of effort.

And most importantly, none of these proposals form a part of a bigger vision.
Instead, these proposals are isolated patches that can be applied in a single design iteration.
These proposals cannibalize each other's use-cases, but do not fully resolve the pain points Ethereum users are
currently suffering from, and they do not fully align with any kind of vision for a Full Native Account Abstraction.

> ## Specification

Reviewer 5: ```suggestion
## Specification
The aforementioned EIPs can be implemented by the following transaction subtypes:
```

enabling Gas Abstracted new smart contract deployment.

### Gas-sponsored time-limited batched execution EOA transaction subtype

```
BATCHED_EOA_SUBTYPE = x
```

This transaction subtype extends the `SPONSORED_EOA_SUBTYPE` transactions with an ability to initialize multiple
EOA calls with a single signature and a single Paymaster validation.

The transaction payload for this subtype is extended with the following fields:
```
 targets
 values
 gasLimits
 atomicity,
 yParity, r, s
```
The `targets` parameter is interpreted as an RPL encoding of `N` destination addressed.
The `values` parameter is interpreted as an RPL encoding of `N` 256-bit amounts.
The `callData` parameter is interpreted as an RPL encoding of `N` call data byte arrays.
The `gasLimits` parameter is interpreted as an RPL encoding of `N` 256-bit gas limit values.

The field `atomicity` contains an RLP encoded array of `N` elements of `byte` "atomicity" values.
Atomicity describes a behaviour in case of a top-level revert in one of the frames, and can take the following values:

* `CONTINUE_ON_REVERT = 0` - ignores the reverted execution and continues executing the batch

> * `TERMINATE_ON_REVERT = 1` - stops the execution of the batch on the reverted execution, does not execute the remainder

Reviewer 5: ```suggestion
* `TERMINATE_ON_REVERT = 1` - stops the execution of the batch on the reverted execution without reverting previous calls in the batch. Does not execute the remainder 
```

Abstraction", or be evaluated on their own merit regardless of the eventual Account Abstraction design.

This proposal is intended to prevent the fragmentation and inconsistencies that may arise as a result of implementing
multiple independent standalone proposals by combining all of them into a unified design.

Here is the list of EIPs that propose adding features to EOA accounts that are addressed by this document:

* [EIP-2711: Sponsored, expiring and batch transactions](https://eips.ethereum.org/EIPS/eip-2711)
* [EIP-3074: AUTH and AUTHCALL opcodes](https://eips.ethereum.org/EIPS/eip-3074)
* [EIP-5003: Insert Code into EOAs with AUTHUSURP](https://eips.ethereum.org/EIPS/eip-5003)
* [EIP-5081: Expirable Transaction](https://eips.ethereum.org/EIPS/eip-5081)
* [EIP-5806: Delegate transaction](https://eips.ethereum.org/EIPS/eip-5806)
* [EIP-7377: Migration Transaction](https://eips.ethereum.org/EIPS/eip-7377)
* [EIP-7553: Separated Payer Transaction](https://eips.ethereum.org/EIPS/eip-7553)

Each one of these proposals introduces features that can benefit both existing users and potential future users of
the Ethereum blockchain in multiple ways.

> However, together these proposals consist of 3 different transaction types and 3 opcodes, which is probably unsustainable.

Reviewer 3: ```suggestion
However, together these proposals consist of 3 different transaction types and 3 opcodes, which is probably unsustainable. The transaction types are incompatible with each other, forcing the user to only use one.
```

```
  chainId,
  nonce,
  callData,
  paymasterData,
  paymasterGasLimit,
  maxPriorityFeePerGas,
  maxFeePerGas,
  validAfter,
  validUntil,
  accessList
```

### Gas-sponsored time-limited EOA transaction subtype

```
SPONSORED_EOA_SUBTYPE = x
```

This RIP-7560 transaction subtype allows ECDSA signature and nonce validation, same as the one used in all
of existing transaction types, to be applied to a transaction that can also provide `paymasterData`, `validAfter`
and `validUntil` parameters.

The transaction payload for this subtype is extended with the following fields:
```
 to
 value
 callGasLimit,
 yParity, r, s
```

This subtype extends the ability of an RIP-7560 transaction to specify a Paymaster contract, 
which accepts the gas charges, to all EOA accounts.
It also provided an EOA with an ability to explicitly specify a time range where this
transaction is valid, creating both "expiring" and "delayed" transactions.

In case the `to` field is `nil`, `callData` is treated as `initCode` for the purpose of contract creation,
enabling Gas Abstracted new smart contract deployment.

### Gas-sponsored time-limited batched execution EOA transaction subtype

```

> BATCHED_EOA_SUBTYPE = x

Reviewer 3: I wonder if this one isn't redundant. Does it offer any advantage over a `DELEGATED_EOA_SUBTYPE` transaction that delegates to a Multicall (or a modified Multicall that supports the three `atomicity` types)?

Author: There is some redundancy in these transaction types of course. Delegated call can be functionally equivalent to a batched call, which is equivalent to a sponsored call for a batch with a single transaction.

I think one issue is a question of a slightly increased 'trust surface'. Delegating control of an EOA to a smart contract means fully trusting this contract, and signing such a transaction will need a new defence level from the wallet apps. Validating a `Multicall` code is simply not something they currently do. Signing a batch, on the other hand, is exactly what they do now - show the payload and ask for confirmation.

Another is a question of EOA's storage. It is not currently defined and there are arguments not to introduce it yet. In this case the EOA's context will be different from other EVM `staticcall` contexts because it cannot use SSTORE itself but can call other contract's code that uses SSTORE.

We can discuss which of the subtypes make most sense and remove the redundant ones.

```

This transaction subtype extends the `SPONSORED_EOA_SUBTYPE` transactions with an ability to initialize multiple
EOA calls with a single signature and a single Paymaster validation.

The transaction payload for this subtype is extended with the following fields:
```
 targets
 values
 gasLimits
 atomicity,
 yParity, r, s
```
The `targets` parameter is interpreted as an RPL encoding of `N` destination addressed.
The `values` parameter is interpreted as an RPL encoding of `N` 256-bit amounts.
The `callData` parameter is interpreted as an RPL encoding of `N` call data byte arrays.

> The `gasLimits` parameter is interpreted as an RPL encoding of `N` 256-bit gas limit values.

Reviewer 3: ```suggestion
The `targets` parameter is interpreted as an RLP encoding of `N` destination addressed.
The `values` parameter is interpreted as an RLP encoding of `N` 256-bit amounts.
The `callData` parameter is interpreted as an RLP encoding of `N` call data byte arrays.
The `gasLimits` parameter is interpreted as an RLP encoding of `N` 256-bit gas limit values.
```

* `ROLLBACK_ON_REVERT = 2` - rolls back all state changes caused by all the execution frames in current transaction

These parameters represent all the inputs necessary to perform `N` sub-transactions.

### Gas-sponsored time-limited one-time delegated EOA transaction subtype

```
DELEGATED_EOA_SUBTYPE = x
```

The transaction payload for this subtype is extended with the following fields:
```
 destination
 callGasLimit,
 yParity, r, s
```

This transaction subtype extends the `SPONSORED_EOA_SUBTYPE` transactions with an ability to execute arbitrary
contract code in the context of the EOA sender.

Note that such transaction provides a one-time signature delegating execution to the `destination` contract,
and this signature cannot be used repeatedly. The private key of the EOA will be needed to sign consequent delegations.

### Gas-sponsored time-limited reusable authorization EOA transaction subtype

```
AUTHORIZED_EOA_SUBTYPE = x
```

The transaction payload for this subtype is extended with the following fields:
```

>  chainId,

Reviewer 3: `chainId` is already in the shared payload that this subtype extends.

The `chainId` should be a part of the signed auth, but if the transaction already has it, we don't need to include it twice.

Author: Fixed.

 nonce,

>  invokerAddress,

Reviewer 3: ```suggestion
 implementationAddress,
```

In the description below you call it `implementationAddress` so let's make it consistent.

>  callData,

Reviewer 3: `callData` is already in the shared payload that this subtype extends.

Author: Fixed.

 callGasLimit
 bigNonce, // TBD: in this transaction we would need both nonces - account state and 2D
 validationGasLimit,
 commit, // 32 bytes hash
 yParity, r, s
```

This transaction subtype extends the `DELEGATED_EOA_SUBTYPE` transactions with an ability to authorize some
contract code to act as a Smart Contract Account and control the context of the EOA sender.

The transaction signature (`yParity`, `r`, `s`) are interpreted as an ECDSA signature on the `secp256k1` curve over
the message `keccak256(MAGIC || chainId || nonce || implementationAddress || commit)`.

The `commit` value is a hash values that the implementation will validate.
It is passed to the validation frame as part of the encoded transaction structure. 

Note that account's `nonce` is not incremented when `AUTHORIZED_EOA_SUBTYPE` transaction is executed so the same
authorization can be used unlimited number of times. In order to withdraw authorization, the account must perform

> any action that will increment its nonce.

Reviewer 3: I suggested this nonce-based revocation technique for 3074 since there was nowhere else to store this information.  But here we also have NonceManager.  Shouldn't we leave `nonce` alone (so that the EOA can transact freely without revoking), and instead revoke authorizations via NonceManager?

The 2D nonce could be `(implementation,seq)`, which gets used as `bigNonce` (which we could rename to something like `delegationNonce` for this transaction subtype), and the user could revoke it by sending any type4 transaction that uses and burns `(implementation,currentSeq)`

I think the EOA's `nonce` then becomes irrelevant for the purpose of delegation.

> Transaction hash uniqueness is therefore guaranteed by a `bigNonce` field for the `NonceManager` contract.

Reviewer 3: The dependency on NonceManager in this subtype means that this RIP will require not just 7560 but also the one in [PR5](https://github.com/eth-infinitism/RIPs/pull/5/files).  This dependency only appears in the `AUTHORIZED_EOA_SUBTYPE` subtype, which may be a reason to move it to its own RIP.  The other EOA related subtypes only require 7560.

Note that most of the transaction fields are not signed by this transaction's signature.
This means that the "implementation" contract is the entity responsible for implementing all kinds of protections,
and the EOA account that signs an authorization fully cedes control to the "implementation" until
the authorization is revoked.

For the purposes of Gas Abstraction, the `paymasterData` parameter is not signed as well,
so different Paymaster contracts can be used for subsequent authorized transactions using the same authorization.

#### Using permanent authorizations to create Smart Contract Accounts

This approach relies on the so-called "Nick's method", which includes deriving an address from some data and
a randomized signature value, without ever knowing the ECDSA private key that would be able to control this address.

Applying this approach to an `AUTHORIZED_EOA_SUBTYPE` transaction, a new account can be created,
whose address is forever bound to `implementationAddress` and `commit`,
without ever needing to deploy code for the contract.

### Gas-sponsored time-limited code injection EOA transaction subtype

```
INSERT_CODE_EOA_SUBTYPE
```

> [//]: # (TODO: 7377 probably needs codeAddr/storage simply because it has no deployment frame; we can have initCode)

Reviewer 3: 7377 did this for optimization.  It does have a call frame (not for deployment but it's executed at the end of the deployment and could set the storage).  It was meant to save gas, and is sustainable because clients only keep each bytecode once in the database and add references to it.  Adding another account that references the same code can therefore be cheaper than CREATE.

Author: So is this something we should support? This is basically making a proxy contract enshrined in the protocol, which includes some changes to account structure and how EVM interacts with "aliased" accounts, so probably would require a separate EIP, wouldn't it?

Reviewer 3: I don't think it requires changes to the account structure, nor should it affect how EVM interacts with these accounts.  In 7377 these are not changed.  It just makes the transition cheaper without becoming a DoS vector, since we know there is already an entry in the bytecodes db and the account can reference it.

OTOH, I don't see this as an important feature.  It's fine to include it as an optimization as long as it doesn't add complexity, but otherwise I'd drop it and use initCode.

Author: @yoavw @drortirosh please see the commit changing the logic from EIP-7377-like to legacy-like:

https://github.com/eth-infinitism/RIPs/pull/1/commits/f568e828858cef36ef73ab21999c0e8f562f77ef

Is this better?

This transaction subtype extends the `SPONSORED_EOA_SUBTYPE` transactions with an ability to insert
contract code in the context of the EOA sender.

The transaction payload for this subtype is extended with the following fields:
```
callGasLimit,
codeAddr,
storage,
data,
value
```

[//]: # (TODO: TBD: this is copied directly from 7377 as it seems to apply well.)

Unlike standard contract deployment, a migration transaction directly specifies what code value the sender's
account should be set to.

As the first step of processing the transaction, set the sender’s code to `state[tx.codeAddr].code`.
Next, for each tuple in `storage` and the sender’s storage trie, set `storage[t.first] = t.second`.

Now instantiate an EVM call into the sender's account using the same rules as EIP-1559 and set the

> transaction's origin to be `keccak256(sender)[0..20]`.

Reviewer 3: > origin to be `keccak256(sender)[0..20]`

Do we want to copy this?  7377 did this to avoid having to deal with the issues raised by letting `tx.origin` have code execution.  The RIP-7560 family already does that anyway.  In addition, the delegation subtype above makes it possible for `tx.origin` to have code (similarly to 3074), so there's no need to have special treatment for ORIGIN in this subtype.

Author: Agreed, removed.

## Rationale

### Authorization as a transaction subtype instead of a family of opcodes

Both approaches have their benefits and downsides.
Here are some of the benefits of a transaction subtype approach:

* Use of `AUTH` and `AUTHCALL` opcodes only allow an unnecessarily dangerous way for EOAs to achieve Gas Abstraction.

    While technically it is possible to implement a gas relaying feature on top of EIP-3074, this approach requires
users to fully hand over control of their accounts to the gas paying "invoker" contract.
A lot can go wrong with this approach.

> 

Reviewer 3: Another benefit that may be worth mentioning, is that it allows smart accounts and EOAs to use the same paymasters.  A dapp that sponsors a transaction doesn't have to care whether the user has a smart account or an EOA.  And paymasters like TokenPaymaster could allow both smart accounts and EOAs to pay with ERC20 tokens.

The 3074-based approach, in contrast, would require developing separate gas abstraction invokers and dapps would have to be aware of the user's account type.

Author: I wanted to mention it but I feel like an adapter layer between ERC-4337 and EIP-3074 is theoretically possible. There can be a `singleton AuthSender` that checks the `AUTH` signature on-chain in `validateUserOp()` and executes an `AUTHCALL` in `execute()`.
This is very inefficient but does technically allow a Paymaster reuse, doesn't it?

Reviewer 3: Yes, I proposed this a while ago and there's already an invoker that does this.  But in the current proposal, gas abstraction doesn't require delegating the EOA to an invoker.  The EOA is just used as an EOA, and may set TransactionType4.paymaster to any paymaster - which only pays for the gas and doesn't even need to know that the account is an EOA.

Author: Added.

* Avoids exposing ECDSA signature and account nonces to the EVM opcodes.
    Using opcodes to verify authorization signature necessarily means that EVM becomes exposed to these details.
    There are real issues with ECDSA signatures, so abstracting cryptography details from EVM opcodes is desired.

* There is only a limited number of available opcodes across all L1 and L2 networks.

    Transaction types are also limited, however with subtypes the namespace is basically unlimited.

    This makes it easier to avoid network splits, as well as to introduce changes to authorisation behaviour if necessary.

## Backwards Compatibility

This RIP suggest creation of new RIP-7560 transaction subtypes, which by itself does not affect any existing contracts.

However, any contract that assumes that an EOA may not have any code or storage, will have its expectations broken
by any of these transaction subtypes. EOAs may be detected by performing an `ecrecover` or by accessing `tx.origin`. 

## Security Considerations

The `DELEGATED_EOA_SUBTYPE` and especially `AUTHORIZED_EOA_SUBTYPE` transactions add an ability for a contract to

> assume control over an EOA. This is a very powerful tool that has very serious security implications.

Reviewer 3: ```suggestion
assume control over an EOA. This is a very powerful tool that has serious security implications.
```

One `very` is enough :)

Wallets must make sure the users know what they are signing and what the implications for the signing will be.
Unlike other similar proposals that can rely on a fact that most wallet do not have an API to sign arbitrary data,
these transaction subtypes could potentially be signed with a regular `eth_signTransaction` API of a wallet

> with RIP-7560 support.

Reviewer 3: For that reason, I wonder if all these subtypes belong in the same RIP, or in a family of RIPs.  An L2 might choose to implement some of the subtypes while excluding others for security reasons.  WDYT?

Author: I agree but I wanted to avoid creating 4 new RIPs.
Should I do it?
The resulting RIPs will be somewhat similar and their 'motivation'/'abstract'/'rationale' sections will be pretty repetitive, only 'specification' will have any new information.

Reviewer 3: Yes, they'll have a lot of similarity but it enables supporting some without supporting others.  I was hoping to make things optional whenever possible.  OTOH it's likely that they'll be adopted together anyway.

In case the signer's account address contains code before `initCode` runs the transaction reverts immediately.

## Rationale

### Authorization as a transaction subtype instead of a family of opcodes

Both approaches have their benefits and downsides.
Here are some of the benefits of a transaction subtype approach:

* Use of `AUTH` and `AUTHCALL` opcodes only allow an unnecessarily dangerous way for EOAs to achieve Gas Abstraction.

    While technically it is possible to implement a gas relaying feature on top of EIP-3074, this approach requires
users to fully hand over control of their accounts to the gas paying "invoker" contract.
A lot can go wrong with this approach.

    Another benefit is that it allows smart accounts and EOAs to use the same paymasters.

    A dapp that sponsors a transaction doesn't have to care whether the user has a smart account or an EOA.
    And paymasters like TokenPaymaster could allow both smart accounts and EOAs to pay with
    [ERC-20](./eip-20) tokens.

    The 3074-based approach, in contrast, would require developing separate gas abstraction invokers and dapps

>     would have to be aware of the user's account type.

Reviewer 4: I don't think this is true. 4337 can be extended to have an invoker which performs the EOA subtype transactions described in this document. If 4337 decides not to change, then this may be true. In which case it should be noted.
